File: frontend/src/services/api/roomService.ts
import apiClient from './authService';
import { Room, RoomRequest } from '../../types/room';

// Room with availability status
export interface RoomWithAvailability extends Room {
    isAvailable: boolean;
}

export const roomService = {
    getAllRooms: async () => {
        const response = await apiClient.get<Room[]>('/v1/rooms');
        return response.data;
    },

    getRoom: async (id: number) => {
        const response = await apiClient.get<Room>(`/v1/rooms/${id}`);
        return response.data;
    },

    // Get rooms with availability status for a specific date and slot
    getRoomAvailability: async (date: string, slotNumber: number): Promise<RoomWithAvailability[]> => {
        const response = await apiClient.get<RoomWithAvailability[]>('/v1/rooms/availability', {
            params: { date, slotNumber }
        });
        return response.data;
    },

    // Get IDs of rooms currently in use based on actual slot times from database
    getCurrentlyInUseRoomIds: async (): Promise<Set<number>> => {
        const now = new Date();
        const date = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-${String(now.getDate()).padStart(2, '0')}`;
        const time = `${String(now.getHours()).padStart(2, '0')}:${String(now.getMinutes()).padStart(2, '0')}`;
        const response = await apiClient.get<number[]>('/v1/rooms/currently-in-use', {
            params: { date, time }
        });
        return new Set(response.data);
    },

    createRoom: async (data: RoomRequest) => {
        const response = await apiClient.post<Room>('/v1/rooms', data);
        return response.data;
    },

    updateRoom: async (id: number, data: RoomRequest) => {
        const response = await apiClient.put<Room>(`/v1/rooms/${id}`, data);
        return response.data;
    },

    deleteRoom: async (id: number) => {
        await apiClient.delete(`/v1/rooms/${id}`);
    }
};

File: backend/src/main/java/com/fams/backend/service/RoomService.java
package com.fams.backend.service;

import com.fams.backend.dto.request.RoomRequest;
import com.fams.backend.dto.response.RoomAvailabilityResponse;
import com.fams.backend.dto.response.RoomResponse;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;
import java.util.Set;

public interface RoomService {
    List<RoomResponse> getAllRooms();

    RoomResponse getRoom(Long id);

    RoomResponse createRoom(RoomRequest request);

    RoomResponse updateRoom(Long id, RoomRequest request);

    void deleteRoom(Long id);

    /**
     * Get all rooms with availability status for a specific date and slot
     */
    List<RoomAvailabilityResponse> getRoomAvailability(LocalDate date, Integer slotNumber);

    /**
     * Get IDs of rooms that are currently in use at the given date+time,
     * using the actual SlotType start/end times from the database.
     */
    Set<Long> getInUseRoomIds(LocalDate date, LocalTime time);
}

